File: z3/added_corner.py
#!/usr/bin/python -u
# -*- coding: latin-1 -*-
# 
# Added corner puzzle in Z3
# 
# Problem from http://www.delphiforfun.org/Programs/AddedCorners.htm
# """
# This puzzle requires that you enter the digits 1 through 8 in the circles and
# squares (one digit in each figure) so that the number in each square is equal
# to the sum on the numbers in the circles which adjoin it.  
# ...
# 
#    C F C
#    F   F
#    C F C
# """
# 
# This Z3 model was written by Hakan Kjellerstrand (hakank@gmail.com)
# See also my Z3 page: http://hakank.org/z3/
# 
from z3_utils_hakank import *

sol = Solver()

n = 8

L = makeIntVector(sol,"L",n,1,n)
[A,B,C,D,E,F,G,H] = L

# constraints
sol.add(Distinct(L))
sol.add(B == A + C)
sol.add(D == A + F)
sol.add(E == C + H)
sol.add(G == F + H)

num_solutions = 0
while sol.check() == sat:
  num_solutions += 1
  mod = sol.model()
  print mod.eval(A), "   ", mod.eval(B), "   ", mod.eval(C)
  print mod.eval(D), "         ", mod.eval(E)
  print mod.eval(F), "   ", mod.eval(H), "   ", mod.eval(H)
  print 
  getDifferentSolution(sol,mod,L)

print "num_solutions", num_solutions  

File: swi_prolog/added_corner.pl
/*

  Added corner puzzle in SWI Prolog

  Problem from http://www.delphiforfun.org/Programs/AddedCorners.htm
  """
  This puzzle requires that you enter the digits 1 through 8 in the circles and 
  squares (one digit in each figure) so that the number in each square is equal 
  to the sum on the numbers in the circles which  adjoin it.  
  ...
  
    C F C
    F   F
    C F C
  """

  Model created by Hakan Kjellerstrand, hakank@gmail.com
  See also my SWI Prolog page: http://www.hakank.org/swi_prolog/

*/

:- use_module(library(clpfd)).
:- use_module(hakank_utils).

go :-
        findall(X, added_corner(X),L),
        length(L,Len),
        format("It was ~d solutions.\n", Len).


added_corner(X) :-
        Digits = 1..8,

        X = [A,B,C,D,E,F,G,H],
        X ins Digits,

        all_different(X),
        B #= A + C,
        D #= A + F,
        E #= C + H,
        G #= F + H,

        label(X),

        format("~d ~d ~d\n", [A,B,C]),
        format("~d   ~d\n",  [D,E]),
        format("~d ~d ~d\n", [F,G,H]),
        nl.

        
